src/java.base/share/classes/sun/security/ssl/SSLConfiguration.java

```diff
@@ -133,6 +138,50 @@ final class SSLConfiguration implements Cloneable {
         useExtendedMasterSecret = supportExtendedMasterSecret;
     }
 
+    static {
+        // jdk.tls.maxCertificateChainLength property has no default
+        Integer certLen = GetIntegerAction.privilegedGetProperty(
+                "jdk.tls.maxCertificateChainLength");
+        if (certLen == null || (int)certLen < 0) {
+            certLen = 0;
+        }
+        maxCertificateChainLength = certLen;
+
+        // Default for jdk.tls.maxClientCertificateChainLength is 8
+        int clientLen = GetIntegerAction.privilegedGetProperty(
+                "jdk.tls.maxClientCertificateChainLength", 8);
+        if (clientLen < 0) {
+            clientLen = 8;
+        }
+
+        // Default for jdk.tls.maxServerCertificateChainLength is 10
+        int serverLen = GetIntegerAction.privilegedGetProperty(
+                "jdk.tls.maxServerCertificateChainLength", 10);
+        if (serverLen < 0) {
+            serverLen = 10;
+        }
+
+        /*
+         * When jdk.tls.maxCertificateChainLength is set, it will override
+         * the defaults for jdk.tls.maxClientCertificateChainLength and
+         * jdk.tls.maxServerCertificateChainLength if they are not explicitly
+         * overridden. That means if either of those properties have been set,
+         * the jdk.tls.maxCertificateChainLength property will not override
+         * the values.
+         */
+        if (maxCertificateChainLength > 0) {
+            if (clientLen == 8) {
```

Comment:
Since 8 is the default for "jdk.tls.maxClientCertificateChainLength", it is going to be overridden when "jdk.tls.maxCertificateChainLength" is set. Setting "jdk.tls.maxClientCertificateChainLength" to 8 is treated as keeping the original default like no-op.

Reply:
`jdk.tls.maxCertificateChainLength` will only override `jdk.tls.maxClientCertificateChainLength` if `jdk.tls.maxCertificateChainLength` is set AND `jdk.tls.maxClientCertificateChainLength` is using the default. For the case your provided here, `jdk.tls.maxClientCertificateChainLength` will be overridden to be 4 which is set by `jdk.tls.maxCertificateChainLength`.

Reply:
Yes, I agree that if the application sets jdk.tls.maxClientCertificateChainLength or jdk.tls.maxServerCertificateChainLength, it should always take precedence even if the specified value is the same as the default. This would also preserve application behavior if one day we want to change the default to something else.

You will need to first see if these properties are set before assigning the default value.

Reply:
Fixed.